src/expr/src/scalar/mod.rs

```diff
@@ -183,6 +183,27 @@ impl MirScalarExpr {
         f(self);
     }
 
+    /// A generalization of `visit_mut`. The function `pre` runs on a
+    /// `MirScalarExpr` before it runs on any of the child `MirScalarExpr`s.
+    /// The function `post` runs on child `MirScalarExpr`s first before the
+    /// parent. Optionally, `pre` can return which child `MirScalarExpr`s, if
+    /// any, should be visited (default is to visit all children).
+    pub fn visit_custom_mut<F1, F2>(&mut self, pre: &mut F1, post: &mut F2)
```

Comment:
Do you have strong feelings about the name? I think `visit_custom_mut` is maybe a bit vague. The only thing I could think of was `visit_mut_pre_post`; what do you think about that?

Reply:
I'm not particularly attached to the name. `visit_mut_pre_post` is all right, though it does not capture the fact it allows you skip going to certain nodes.